### ISSUE TITLE
Radial menu is invisible/unusable sometimes (BUG)

### ISSUE THREAD

**Issue by InsaneGrox** [Frooxius/NeosPublic#977]
Sometimes the radial menu fails to load or initialize, or is invisible, happens to both me and my friends on occasion and is fustrating when it happens (buttons still work, it's just we can't see them).

**Comment by shiftyscales** [comment]
Do you have any specific examples of when/how this occurs?

**Comment by Frooxius** [comment]
- Are there any conditions when this happens?
- Does it persist for the entire duration of the session? Or do you have to restart Neos for it to clear up?
- Any logs from such situations?

**Comment by Zyro1331** [comment]
I have this issue too, it occurs randomly. It happens when you lag, or the host is lagging. Like the assets haven't loaded yet in a way. It occurs for like 30 - 15 seconds before you can actually use the menu.

**Comment by H3BO3** [comment]
Doesn't the host have to load your avatar and set up everything in user root? If so it may be related to that

**Comment by Frooxius** [comment]
@Zyro1331 That might be a different issue. If the menu is momentarily unusable, it's typically because the worker threads are too busy and things are congested. It clears up eventually after everything in the queue is handled. It's not specific to these menus. The host lagging shouldn't matter, the menu logic runs on your end locally.

@H3BO3 That won't be the cause. The menus are setup separately and instantly when the user is generated, independently of the actual avatar.

**Comment by Toxic-Cookie** [comment]
It happens to me as well. It seems to happen randomly and isn't reliably reproducible.

**Comment by InsaneGrox** [comment]
About all I can say is that it occurs often when joining sessions with more than 6 players, however I should also mention that it never loads after joining said session (restart is required if it doesn't load in my case)

**Comment by sirkitree** [comment]
I don't have much useful information to add except to say I had this happen to me just once the other day as well. I had been in the world for a while, working with others, and this happened. The menu was there as the raycaster from my hand would collide with it, but I could not actually see the menu or it's items. I moved away from the immediate area and it came back, so it felt like a rendering problem specific to the exact location or perhaps the things that were around me. I had H3BO's Multitool equipped at the time, and was working with logix.

**Comment by H3BO3** [comment]
I had this happen to me yesterday when I loaded into Enverex's space world. As most people know this is a very heavy world and there are a lot of colliders that I assume are being generated on initial load, not to mention the number of assets - During this time my entire UI got held up to the point of being unusable, including the context menus (Worker thread issues I assume? Is it possible very high priority could be given to local UI?). When I hovered my laser over things in the dash I still got haptic events when moving to individual buttons but no visual hover feedback, which I thought was strange. When I pressed my context menu button it seemed to be somewhere else entirely in the world (this was shortly after I got teleported to a different location) and my laser was stretching out sideways in a weird angle.

**Comment by shiftyscales** [comment]
Frooxius has spoken on this prior, @H3BO3:
> They're already prioritized. It's a lot more complicated than that though. If too many prioritized jobs end up causing contention, then it ends up not mattering too much.
> 
> The scheduling of the threads is also handled by the OS, rather than Neos, so some weird things happen as well. Sometimes you end up getting priority inversion, because the higher priority threads are waiting on some lower priority task to finished, but that task is not getting done because the higher priority tasks are taking up all the resources.
> 
> I do want to tweak and fine tune the system a fair bit, but it's not as simple as that unfortunately. At some point you'll get things freezing up anyways, because the system becomes congested with too much work.
> 
> Like I said it's not that simple. The UI isn't necessarily dependent on particular tasks, but on certain system resources. Many of the resources are shared among the higher and lower priority threads and can become a point of contention.
> 
> So technically you'd have to prioritize all the tasks, at which point none of them really has high priority anymore, since they're all the same level again and depend on the OS scheduling.
> 
> Sometimes the prioritization ends up doing worse reasons for that, since it messes with how the OS schedules the threads. It's a really complex issue and it just takes a lot of fine tuning and balancing and replacing systems so they try to avoid contention as much as possible and instead use more concurrent mechanisms. 
> 
> I do want to replace a bunch of the resources with mechanisms designed for high concurrency. Originally Neos used them, but problem with them is that they trade off the high concurrency for some memory allocations. With good GC it's not an issue, but with the one Unity has it ends up having catastrophic impact on performance, causing GC stutters, which is why I replaced them a while back.
> 
> With the incremental GC it should be better though so I'll start playing with that again.

What needs to be done in this case instead is to find a replicable scenario in which this issue consistently occurs- then it can be profiled, and the underlying issue/task that is holding up the UI generation can be tweaked/addressed.

**Comment by Frooxius** [comment]
I believe this is likely the same issue as here: https://github.com/Frooxius/NeosPublic/issues/1111 Closing this for now, since this should help. Reopen if there are more cases of this.

**Comment by Hayden-da-Serg** [comment]
I still get this almost every time I use Neos, as my mobile CPU does not have fun when intense operations are happening, I tend to have to sit there for at least a minute in some cases before any of my UI starts to decide to update (dash, context menus, legacy inventory and so on)
I also see mesh updates running into the same issues.

**Comment by InsaneGrox** [comment]
It still happens, though from what I can gather it's not easily fixable as it's an issue with the UI not loading first, the dash is also affected as it turns out, I do have an idea for a fix/workaround though, don't re-load the UI every time you join a world, keep it in memory from when you first start up neos.

**Comment by Toxic-Cookie** [comment]
Can confirm it never stopped happening to me as well. No idea how to reproduce it other than wait for intense lag to ensue. I'm susing background threads getting clogged up by who knows what.

**Comment by InsaneGrox** [comment]
I'm gonna bring up that this issue seems to apply to the entire goddamn UI and makes Neos unplayable at times...

**Comment by Frooxius** [comment]
I'm sorry to hear that it's still happening, but unfortunately we'll need more information to narrow down the cause.
- Does this happen in specific scenarios? E.g. under heavy load and such?
- Is there a way to reliably reproduce this?
- During this happening to you, can you repeatedly mash the F8 key (make sure Neos window is focused) on your keyboard and send us the logs? This will capture state of background threads and can help identify issues that might be locking those up, we have fixed a few of those in the past, but there can be numerous causes based on what's going on.

**Comment by ProbablePrime** [comment]
This has been happening to me too, now come to think of it. 

Neos locks up sometimes when opening default Context Menus and I see white squares instead of things like tool icons for things like the Dev Tip etc.

I'll get some more information when I head onto Neos today.

**Comment by Frooxius** [comment]
It does sound like the background threads are being clogged with something. The F8 mashing could help gather a lot of useful info there.

I've got one more question that can help too:
- When this happens, how long does this typically last before the UI resumes?

**Comment by InsaneGrox** [comment]
> It does sound like the background threads are being clogged with something. The F8 mashing could help gather a lot of useful info there.
> 
> I've got one more question that can help too:
> 
> * When this happens, how long does this typically last before the UI resumes?

Varies quite a lot... It can vary from less than a second to until restarting the game (by ending it with the task manager because quitting by normal means is impossible)
usually a few seconds though.

**Comment by Frooxius** [comment]
I see.

If it's just a few seconds that's somewhat expected, as that can happen under heavy load and there's not enough system resources available to keep updating everything. That's mostly down to having more optimizations and doing some more profiling. 

If it takes a while then something has probably heavily locked up the processes. Mashing the F8 key in those cases and sending a log can help sample some data on what's happening.

**Comment by ProbablePrime** [comment]
@Frooxius did you mean F9?

**Comment by shiftyscales** [comment]
> @Frooxius did you mean F9?

Yes.
```
- Pressing F9 on keyboard will log a snapshot of the background workers to the log file
-- Added to help diagnose some resource starving issues (reported by @Rukio, @ProbablePrime | Docs and @H3BO3)
-- If you experience UI freezing, press the F9 key repeatedly (about twice a second is enough) for about 10-20 seconds and then send the logs
```
https://discord.com/channels/402159838827905024/469131434628612116/736124248065572886
